DOC: Rewritten style.py docstrings: #12148

aaronmichaelsilberstein opened the pull request:
Docstrings for pandas/io/formats/style.py now accurately reflects the behavior of the ``Styler`` class. Fixed a typo in the docstring of the `style` property of ``DataFrame``.

- [x] closes #12148 
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

aaronmichaelsilberstein committed:
DOC: Rewritten style.py heredoc documentation: #12148

Heredoc documentation in pandas/io/formats/style.py now accurately reflects the behavior of the ``Styler`` class. Fixed a typo in the heredoc of the `style` property of ``DataFrame``.
Closes #12148

pep8speaks commented:
Hello @aaronmichaelsilberstein! Thanks for submitting the PR.



 - In the file [`pandas/io/formats/style.py`](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py), following are the PEP8 issues :

> [Line 176:17](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L176): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 177:17](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L177): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 178:17](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L178): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 179:17](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L179): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 194:45](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L194): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 195:45](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L195): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 196:45](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L196): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 197:45](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L197): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 198:45](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L198): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 199:45](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L199): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 674:56](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L674): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 855:39](https://github.com/pandas-dev/pandas/blob/94a6dcbcf5091c85a873b20a8cf8ff697306f0bd/pandas/io/formats/style.py#L855): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

aaronmichaelsilberstein committed:
Fixed PEP8 issues

TomAugspurger commented on pandas/io/formats/style.py:
```diff
@@ -49,19 +48,21 @@ def _mpl(func):
     """
-    Helps style a DataFrame or Series according to the
-    data with HTML and CSS.
+    A decorator class for a `DataFrame` object to provide an
```
What's a "decorator class"? This like it'd be easily confused with python decorators, which I don't think you mean.

aaronmichaelsilberstein replied:
By "decorator class" I mean the concept precisely defined on p. 175 of "Design Patterns" by Gamma et al (commonly referred to as Gang of Four, or GOF), which seems to precisely describe what Styler is: a class which contains additional state information to perform an auxiliary function on another class.  Is there a name we consistently use in our documentation to refer to such a concept?

TomAugspurger commented on pandas/io/formats/style.py:
```diff
@@ -72,32 +73,46 @@ class Styler(object):
-
-    * Blank cells include ``blank``
-    * Data cells include ``data``
+    `Styler` renders HTML on a `DataFrame` using its own `Styler.render`
```
I don't follow "renders HTML on a DataFrame". Perhaps "renders an HTML representation of a DataFrame"?

TomAugspurger commented on pandas/io/formats/style.py:
```diff
@@ -72,32 +73,46 @@ class Styler(object):
+    independent from, its `data` object's `DataFrame.to_html` rendering
+    method.
+
+    `Styler.render` returns an HTML string, based on attributes which are set
```
Don't need this comma.

TomAugspurger commented on pandas/io/formats/style.py:
```diff
@@ -72,32 +73,46 @@ class Styler(object):
-    return values with strings containing CSS ``'attr: value'`` that will
-    be applied to the indicated cells.
 
-    If using in the Jupyter notebook, Styler has defined a ``_repr_html_``
```
Why did you remove this? I think it's helpful to note that the class defines `_repr_html_`.

aaronmichaelsilberstein replied:
I didn't remove it - it's now in line 109.

TomAugspurger commented on pandas/io/formats/style.py:
```diff
@@ -764,22 +833,40 @@ def set_caption(self, caption):
+        -----
+        Entries are rendered with CSS selectors according to gross properties
+        of their position and content:
+        * Index and Column names include ``index_name`` and ``level<k>``
```
`set_table_styles` doesn't seem like the appropriate place for this section. Perhaps `.render`?

TomAugspurger commented on pandas/io/formats/style.py:
```diff
@@ -813,7 +900,7 @@ def hide_columns(self, subset):
         Parameters
         ----------
         subset: IndexSlice
-            An argument to ``DataFrame.loc`` that identifies which columns
```
@jorisvandenbossche Do we have an official policy on single vs. double backticks? I know numpydoc recommends single, but I thought we used double so that it looked nicer in the HTML docs.

TomAugspurger commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -250,8 +250,7 @@ Performance Improvements
 ~~~~~~~~~~~~~~~~~~~~~
-
--
+- Heredoc documentation for io/formats/style.py has been overhauled to more accurately reflect the functioning of the `Styler` class. (:issue:`12148`)
```
What's a "Heredoc"?

TomAugspurger commented:
Thanks, looks like a nice improvement. Just a few questions.

aaronmichaelsilberstein committed:
Changed 'heredoc' to 'docstring'

aaronmichaelsilberstein committed:
Code review modifications

Got rid of comma.
Changed "renders HTML on a DataFrame" to "renders an HTML representation of a `DataFrame`".
Moved the notes for .set_table_values to the notes for .render, and updated the notion of semi-data-dependent properties in the class docstring.

aaronmichaelsilberstein reviewed (commented):
Thanks for your careful review!

aaronmichaelsilberstein committed:
Fixed iPython docstring example for .format

jreback commented:
closing as stale. ping if you want to update.
